### PR TITLE
luminous: mgr: drop GIL in get_config

### PR DESCRIPTION
**Pull request by badone**
https://tracker.ceph.com/issues/38460

### CODE REVIEW

**Comment by smithfarm** [comment]
Same "Conflicts" section nitpick as in the mimic PR.

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -420,15 +420,13 @@ void ActivePyModules::notify_all(const LogEntry &log_entry)
...
-  PyThreadState *tstate = PyEval_SaveThread();
-  Mutex::Locker l(lock);
-  PyEval_RestoreThread(tstate);
-
   const std::string global_key = PyModuleRegistry::config_prefix
     + module_name + "/" + key;
 
   dout(4) << __func__ << "key: " << global_key << dendl;
 
+  Mutex::Locker l(lock);
```
Comment (smithfarm) [inline src/mgr/ActivePyModules.cc]: mimic has `Mutex::Locker lock(module_config.lock);` here - is `l(lock)` the luminous version of that?

Comment (badone) [reply src/mgr/ActivePyModules.cc]: @smithfarm In luminous PyModuleConfig (the type of module_config (post luminous only) and config_cache) is a std::map and does not have its own lock. Post luminous PyModuleConfig was converted to a class object with its own lock so we take the more global "ActivePyModules::lock" here.

**Comment by yuriw** [comment]
[wip-yuri6-testing-2019-03-11-1637-luminous](https://trello.com/c/XFEKYs8N/554-wip-yuri6-testing-2019-03-11-1637-luminous)

**Comment by yuriw** [comment]
@badone @smithfarm This passed tests and @neha-ojha is OK with the change.
Pls approve and feel free to merge

**Comment by wumingqiao** [comment]
It can NOT resolve https://tracker.ceph.com/issues/38460.
I have merge it into my branch, and the standby mgr deadlock again:
[root@hci1 ceph]# pstack 2535
...
Thread 11 (Thread 0x7f01b45da700 (LWP 2802)):
#0  0x00007f01bb020a0b in do_futex_wait.constprop.1 () from /lib64/libpthread.so.0
#1  0x00007f01bb020a9f in __new_sem_wait_slow.constprop.0 () from /lib64/libpthread.so.0
#2  0x00007f01bb020b3b in sem_wait@@GLIBC_2.2.5 () from /lib64/libpthread.so.0
#3  0x00007f01bcfab5f5 in PyThread_acquire_lock () from /lib64/libpython2.7.so.1.0
#4  0x00007f01bcf77156 in PyEval_RestoreThread () from /lib64/libpython2.7.so.1.0
#5  0x00007f01bcfaf2ee in lock_PyThread_acquire_lock () from /lib64/libpython2.7.so.1.0
#6  0x00007f01bcf7ebb0 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#7  0x00007f01bcf80efd in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#8  0x00007f01bcf7e3fc in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#9  0x00007f01bcf7e57d in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#10 0x00007f01bcf7e57d in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#11 0x00007f01bcf7e57d in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#12 0x00007f01bcf80efd in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#13 0x00007f01bcf0a94d in function_call () from /lib64/libpython2.7.so.1.0
#14 0x00007f01bcee59a3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#15 0x00007f01bcf795bd in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#16 0x00007f01bcf80efd in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#17 0x00007f01bcf0a858 in function_call () from /lib64/libpython2.7.so.1.0
#18 0x00007f01bcee59a3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#19 0x00007f01bcef4995 in instancemethod_call () from /lib64/libpython2.7.so.1.0
#20 0x00007f01bcee59a3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#21 0x00007f01bcf3c947 in slot_tp_init () from /lib64/libpython2.7.so.1.0
#22 0x00007f01bcf3b65f in type_call () from /lib64/libpython2.7.so.1.0
#23 0x00007f01bcee59a3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#24 0x00007f01bcf777b7 in PyEval_CallObjectWithKeywords () from /lib64/libpython2.7.so.1.0
#25 0x000055c05ee05170 in StandbyPyModule::load() ()
#26 0x000055c05ee062ef in StandbyPyModules::start_one(std::string const&, _object*, SafeThreadState const&) ()
#27 0x000055c05ee09fb5 in PyModuleRegistry::standby_start(MonClient*) ()
#28 0x000055c05ee1ef36 in MgrStandby::handle_mgr_map(MMgrMap*) ()
#29 0x000055c05ee1f7f4 in MgrStandby::ms_dispatch(Message*) ()
#30 0x000055c05f229a22 in DispatchQueue::entry() ()
#31 0x000055c05f0070bd in DispatchQueue::DispatchThread::entry() ()
#32 0x00007f01bb01ae25 in start_thread () from /lib64/libpthread.so.0
#33 0x00007f01ba0fb34d in clone () from /lib64/libc.so.6
...
Thread 2 (Thread 0x7f01988a5700 (LWP 13263)):
#0  0x00007f01bb01e945 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x000055c05ee07afe in Cond::Wait(Mutex&) ()
#2  0x000055c05ee05795 in StandbyPyModule::get_config(std::string const&, std::string*) const ()
#3  0x000055c05ee18781 in ceph_config_get(BaseMgrStandbyModule*, _object*) ()
#4  0x00007f01bcf7ebb0 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#5  0x00007f01bcf80efd in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#6  0x00007f01bcf7e3fc in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#7  0x00007f01bcf80efd in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#8  0x00007f01bcf7e3fc in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#9  0x00007f01bcf80efd in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#10 0x00007f01bcf0a858 in function_call () from /lib64/libpython2.7.so.1.0
#11 0x00007f01bcee59a3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#12 0x00007f01bcef4995 in instancemethod_call () from /lib64/libpython2.7.so.1.0
#13 0x00007f01bcee59a3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#14 0x00007f01bcee5a85 in call_function_tail () from /lib64/libpython2.7.so.1.0
#15 0x00007f01bcee5dbb in PyObject_CallMethod () from /lib64/libpython2.7.so.1.0
#16 0x000055c05ee0ccdc in PyModuleRunner::serve() ()
#17 0x000055c05ee0d35f in PyModuleRunner::PyModuleRunnerThread::entry() ()
#18 0x00007f01bb01ae25 in start_thread () from /lib64/libpthread.so.0
#19 0x00007f01ba0fb34d in clone () from /lib64/libc.so.6

The issue is at
`bool StandbyPyModule::get_config(const std::string &key,
                                 std::string *value) const
{
  PyThreadState *tstate = PyEval_SaveThread();
  PyEval_RestoreThread(tstate);

  const std::string global_key = PyModuleRegistry::config_prefix
    + module_name + "/" + key;

  dout(4) << __func__ << "key: " << global_key << dendl;

  return state.with_config([global_key, value](const PyModuleConfig &config){
    if (config.count(global_key)) {
      *value = config.at(global_key);
      return true;
    } else {
      return false;
    }
  });
}`

we should drop GIL in it.

**Comment by badone** [comment]
@wumingqiao yes, this code is racy and https://tracker.ceph.com/issues/38467 is a placeholder for work to clean it up but a closed luminous PR is not the place to discuss this. Please either submit a PR, or open a tracker, or preferably both.

**Pull request merged**